File: Algorithms/Sorting/Radix Sort/RadixSort.cpp
#include<iostream>
using namespace std;

int max(int array[], int N){
    int max = array[0];
    for(int i = 0; i<N;i++){
        if (i>max){
            max = i;
        }
    }
    return max;
}

void countsort(int array[], int N, int e){
    int nn = sizeof(array)/sizeof(array[0]); 
    int output[5];
    int count[10] = {0};

    for(int i=0; i<N;i++){
        count[(array[i]/e) % 10] += 1;
    }

    for(int i=1; i<10;i++){
        count[i] += count[i-1];
    }

    cout << "M: ";
    for(int i=0; i<10;i++){
        cout << count[i] << ',';
    }
    cout << endl;

    for(int i=N-1; i>=0;i--){
        output[count[(array[i]/e) % 10] - 1] = array[i];
        count[(array[i]/e) % 10] -= 1;
    }

    for(int i=0; i<N;i++){
        array[i] = output[i];
    }
}

void radixsort(int array[], int N){
    int m = max(array, N);

    for(int e=1; m/e>1; e*=10){
         countsort(array, N, e);
    }
}


int main(){
    int array[5] = {0,0,0,0,0};

    for(int i = 0; i<(sizeof(array)/sizeof(array[0])); i++){
        cin >> array[i]; 
    }

    cout << "Before: ";
    for(int i: array){
        cout << i << ',';
    }
    cout << endl;

    radixsort(array, sizeof(array)/sizeof(array[0]));

    cout << "After: ";
    for(int i: array){
        cout << i << ',';
    }
    cout << endl;

    return 0;
}

File: Algorithms/Sorting/Radix Sort/RadixSort.py

def countsort(array, N, e):

    output = [0] * 10
    count = [0] * 10
    for i in array:
        count[int(i/e)%10] += 1
    print(count)
    for i in range(1, len(count)):
        count[i] += count[i-1]
    print(count)

    for i in range(len(array)-1, -1, -1):
        output[count[int(array[i]/e)%10]-1] = array[i]
        count[int(array[i]/e)%10] -= 1
    
    for i in range(len(array)):
        array[i] = output[i]

    print(count)


def radixsort(array, N):
    m = max(array)
    
    e = 1
    while int(m/e) > 0 :
        countsort(array, N, e)
        e = e*10

array = []

while(len(array)<5):
    array.append(int(input("Enter your no: ")))
    

print("Before sorting: ",array)

radixsort(array, len(array)-1)

print("After sorting: ",array)